src/status_im2/subs/root.cljs

```diff
@@ -232,6 +232,7 @@
 
 ;; communities
 
+(reg-root-key-sub :my-pending-requests-to-join :my-pending-requests-to-join)
```

Comment:
so for this we need to get a list of users requests to join different communities from backend.

Was not sure if it is better to rename this to `:my-pending-requests-to-join-communities`.

Or maybe better
```
:my-requests {
   :communities {}
   :some-other-section
}
```

Does anyone know if we will have something similar with active requests that need approval? I guess there is this with contacts etc.

Reply:
No right or wrong, but I would generally expect community state to live under a root `:communities` key (everything community related actually).

Additionally, avoiding new state at the root of the app db can help us reduce the overhead in the re-frame *signal graph* because it allows us to design the state with more layer-3 subs (which are cached). Our app could probably get a meaningful boost in performance if we started to implement this practice more aggressively.

Reply:
In general, I think the overall recommendation is to avoid nesting, and prefer "flat" structures rather than nesting.

The suggestion is to structure it similarly to database tables, so in this case you'd have a root key 
`:communities-requests`, and then how you store the data within the map is contextual to how you query it, since you are building up your own "index" effectively (i.e if you query always by community-id `{:communities-requests {:community-id {:request-id-a {}}}}` might be what you want, if you always want the full list, `{:communities-request {:request-id-a {}}` is probably a better option, and there's more I guess.

This is suggested by Mike Thompson https://groups.google.com/g/clojurescript/c/x5RHF0710CU/m/Th1fizgnlmkJ where he mentions that's how it conceptualizes the app db.

When it comes to performance, it's harder to judge for me, since it's dependent on how the data changes and how you nest things I guess, and as Icaro suggested having some level of nesting, could save performance since you are reducing the level-2 subscriptions, and increasing level-3 subs and therefore leveraging caching, but really hard to judge for me in this case, it'd be super interesting to try things out and see, though I have the feeling that it's very contextual to how the data is updated in the database.

What I can say though for sure, is that getting a nested structure right, is more prone to error, and historically performance issues where due mostly to coding error or not being careful with how we placed things, rather than one approach being more performant than the other.
 
For example we had `{:chats {:chat-id-1 {:messages {:message-id-1 {}}}` in the past, which can look natural, but normally will lead to bad performance since messages are updated frequently, and a new message lead to a chat re-render since we were not so careful when creating a subscription that was pulling the chat object which included messages (we should have subscribed to less state to avoid this, but it's an easy mistake to make which would be avoided if we "flatten" the collection).

So I'd err in this case on the side of caution and keep things flatter, mostly because it's less error prone and easier to reason about, although it could be potentially less performant, but don't have a strong opinion on it

Reply:
>In general, I think the overall recommendation is to avoid nesting, and prefer "flat" structures rather than nesting.
>
>The suggestion is to structure it similarly to database tables

I don't see it that way. I see the app db for what it is, an open map, without any of the capabilities of a database. In fact, it's much more similar to the wild west of document-oriented databases like Firestore. But any analogy with a database is very frail IMO (unless we use a proper tool, see below).

>This is suggested by Mike Thompson
>https://groups.google.com/g/clojurescript/c/x5RHF0710CU/m/Th1fizgnlmkJ where he
>mentions that's how it conceptualizes the app db.

I respectfully disagree with him, and I think his comment is disconnected from the `status-mobile` reality. I also didn't suggest deep nesting in any way in my previous comment, which his comment focuses a lot. We might have a different definition of what "deep" means in data structures.

I think we should design the app state around the app's read & write patterns, plus the correct data structures. I believe that's what you're hinting when you say "how you store the data within the map is contextual to how you query it". I completely agree with that. We get that right and a lot of good things naturally follow, but I think data structure flatness is tangential to that.

Going back to the app db being a database. Many products went beyond the conceptualization and actually manage the app state with an in-memory database running on the client, with full support for indexing, query language, etc. This might sound crazy at first, but it's really not. I know of some projects that use [DataScript](https://github.com/tonsky/datascript) (and maybe [re-posh](https://github.com/denistakeda/re-posh)), one famous one is http://logseq.com/. Of course, this would require careful experimentation because every tool brings incidental complexity. Performance wise, it could actually improve, but impossible to say without measurements.

>So I'd err in this case on the side of caution and keep things flatter, mostly because it's less error prone and easier to reason about

I would actually caution against using flatness as a meaningful guide to design the app state. The state should mostly serve the purposes of helping efficiently render trees of components. There are always ways to create abstractions to simplify how the data is read/written (I'm again hinting on the idea of a proper client side database).

But good points overall, when in doubt, let's bring the Occam's razor and keep things simple as you suggest Andrea.